### PR TITLE
chore: uploader can append

### PR DESCRIPTION
**Pull request by fracasula**
# Description

We can append if:
* source is not an ETL source
* source it's not a replay
* source category is not in a specific set
* we are in append mode

### Additional changes
* removed duplicated mocked uploader in favour of MockGen

### TODOs
* use `Uploader.CanAppend()` in the other integrations that support `APPEND` mode

## Linear Ticket

< [Linear Link](https://linear.app/rudderstack/issue/PIPE-104/always-do-merge-if-source-type-is-replayetl-sources-%5Bwarehouse%5D) >

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Comment by fracasula** [comment]
Blocked by https://github.com/rudderlabs/rudder-go-kit/pull/108

File: warehouse/utils/utils.go
Comment (achettyiitr) [inline warehouse/utils/utils.go:233]: Can we move this in a separate package?

Comment (fracasula) [reply warehouse/utils/utils.go:233]: I had created already a dedicated package for it, namely `warehouse/internal/mocks/utils/mock_uploader.go`. I was planning to have other mocks there as well. What's wrong with it?

Comment (fracasula) [reply warehouse/utils/utils.go:233]: So what's your suggestion? What's the exact package name that you'd create for hosting this mock?

Comment (achettyiitr) [reply warehouse/utils/utils.go:233]: The mock package name is fine, I was simply asking if we can move this interface to a separate package `warehouse/uploader`?

File: services/pgnotifier/pgnotifier.go
Comment (achettyiitr) [inline services/pgnotifier/pgnotifier.go:114]: Good catch.

Comment (achettyiitr) [reply services/pgnotifier/pgnotifier.go:114]: Did we get this while running the test with `-race`?

**Review by achettyiitr — APPROVED** [review]

File: services/pgnotifier/pgnotifier.go
Comment (achettyiitr) [reply services/pgnotifier/pgnotifier.go:114]: This has already been take care of with the new changes. These globals have been [removed](https://github.com/rudderlabs/rudder-server/pull/3784/files#diff-81c88462ea98fe7900948848a9b8e201fd33f6efdf9c6770de3cbd64e657bab9R64-R65).

Comment (fracasula) [reply services/pgnotifier/pgnotifier.go:114]: Yes the `-race` highlighted it :+1: I'll try to have a look at that PR too then so maybe we can speed this up. With the atomic change I'm consistently getting failing tests now.

File: warehouse/utils/utils.go
Comment (fracasula) [reply warehouse/utils/utils.go:233]: Right, let's move it as part of another cleanup PR where we refactor this a bit. There are several things we want to address, splitting the interface in a more SOLID approach, move the derivated interfaces in appropriate packages and remove the circular references. We can do it all together as a separate change, wdyt?

Comment (achettyiitr) [reply warehouse/utils/utils.go:233]: Sure.

File: services/pgnotifier/pgnotifier.go
Comment (fracasula) [reply services/pgnotifier/pgnotifier.go:114]: > This has already been take care of with the new changes. These globals have been [removed](https://github.com/rudderlabs/rudder-server/pull/3784/files#diff-81c88462ea98fe7900948848a9b8e201fd33f6efdf9c6770de3cbd64e657bab9R64-R65).

@achettyiitr the race can still happen in your branch as well. if the variable is hot-reloaded while we read then you get a race. it's just going to be harder to see this race in tests because the variable is not global anymore.

Comment (achettyiitr) [reply services/pgnotifier/pgnotifier.go:114]: Understood. Hot reloadable then need to concurrent safe then

**Comment by fracasula** [comment]
@lvrach this one can potentially be merged if we temporarily remove the `-race` from the `Makefile`. I still need another approval though.

**Pull request merged**
